### DISCUSSION TITLE
Is it okay if I add as much tests as possible ?

### DISCUSSION THREAD

**Discussion by matiaslauriti** [eduayme/RescueApp#225] [General]
Hi guys !

I found this project searching on [here](https://up-for-grabs.net/#/filters?tags=laravel).

I see it is almost complete and is really a good project so I would like to participate on it. I have 4 years experience with Laravel and 8+ years experience in this field, I have started with 5.4 and I currently use 8.x, so I think I can help with it in mostly any aspect.

I also know Vue and React, so if you need any feature like that, I could have a look at it.

I also see the project needs a lot of standardizations like:
* Models should go in `App\Models`.
* There is no need to define `$primaryKey` if it is `id`.
* There is no need to define table names when the framework already does this.
* `.idea` or any IDE folder should be `.gitignore`d.
* Personally, adding any DB foreign key, will only create problems, only that should be 100% code checked instead of DB.
* Dev docker should not auto-start, if I have a lot of projects with same ports, it will mess everything all the time, you can make use of `.env` to give what the user wants or use a default, as the [official documentation](https://docs.docker.com/compose/environment-variables/) shows.
* You should be using `FormRequests` instead of plain `Validator`s. It will make the code more readable.
* It would be awesome to upgrade how tests are done, there are some wrong things there and it would be great to use same DB Service (MySQL) but another DB just for testing, instead of SqlLite.
* The most `use` you can write on your code, the better. It is better to do `use App\Model` than `someFunction('App\Model')`, as it will be not trackable and you will really not see how much you are using. Having `use`s at the top, will help see what you are requiring for that class to work.

These are some of the things I saw rapidly... **I am, in no way, saying that what it is already achieved, is wrong, bad or you don't know anything, I am just giving my honest opinion for big projects after all my experience, so please, take this as constructive 🙏**

It would also be awesome to upgrade to Laravel 8 and PHP 8.

If you give me the green light, I will add as much tests as possible and edit current ones.

Thank you !

**Comment by eduayme** [comment]
Hi @matiaslauriti!

First of all thanks to spend your time writing here (the first one!) and give all this feedback about the project.

I started this project more than 2 years ago when I was still in my bachelors and I had no real experience programming. I am aware that there are a lot of problems in this project: lack of standardization, no testing, bad documentation, etc. Of course I am open to all kind of improves experienced people can give to this project. I will review all your points and try to learn from them. Right now, working full time and studying a Masters I have not much time for this but in three weeks I will finish it and start working really hard on this open source project.

For all the reasons you mentioned and other existing problems I am considering rebuild the project for the v2.0. I want to make the frontend and backend independent, build the front with Vue and the backend with Django, improve all code documentation to be easy to contribute to the project, add end-to-end testing with Cypress and more cool things. If any suggestion or comments I am more than open to learn from everyone! 

Thank you, I hope we can all learn from each other

**Reply by matiaslauriti** [reply]
Thank you for taking the time to reply !

I tried to write the original comment as constructive as possible but it still looks a little harsh to my eyes, sorry for that.

I totally understand what you are saying, I have a sort of OCD where I need to have everything as good as possible (specially standards) so I get triggered easy with simple things 😁

So, if you are planning on changing this to Django, I will not be able to contribute as I do not know it, and I do not work with it. I am only with Laravel and VueJs/ReactJs.

Still, really cool project !

**Reply by eduayme** [reply]
No worries, I completely understand it.

The planning on changing to Django is more long term, probably next year, so it would be great to have any change propose in short term. I am still working on some bugs and problems so any MR to refactor things is more that welcome. It would be really helpful to learn my Laravel skills and improve the project so you are more than welcome! 

Thanks for your time and your nice words :)
